keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,739 @@
+  bucketAvailable: [12]
+```
+
+1. `name`: when created by COSI, the `Bucket` name is generated in this format: _"br-"+uuid_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster.
+1. `labels`: added by COSI. The "cosi.io/provisioner" key's value is the provisioner name. The "cosi.io/bucket-prefix" key's value is the `BucketRequest.bucketPrefix` value when specified. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by COSI to defer `Bucket` deletion until backend deletion ops succeed. Implementation may add one finalizer for the BR and one for the BA.
+1. `provisioner`: (optional) The provisioner field as defined in the `BucketClass`. If empty then there is no driver/sidecar, thus the admin had to create the `Bucket` and `BucketAccess` instances.
+1. `retentionPolicy`: Prescribes the outcome when the `BucketRequest` is deleted.
+   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
+   - _Delete_: the bucket and its contents are expected to be deleted by the backend store.
+A `Bucket` is not deleted if it is bound to a `BucketRequest`.
+1. `anonymousAccessMode`: a string specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the bucket objects are intended to be publicly readable and/or writable. One of:
+   - "private": Default, disallow uncredentialed access to the backend storage.
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+   > Note: `anonymousAccessMode` is intended for workloads to consume the bucket out of band, i.e. COSI does not provide a automated mechanism for a k8s workload to consume the bucket anonymously
+1. `bucketClassName`: (set by COSI for greenfield, and ignored for brownfield) Name of the associated bucket class.
+1. `bucketRequest`: (set by COSI for greenfield, and ignored for brownfield) an `objectReference` containing the name, namespace and UID of the associated `BucketRequest`.
+1. `allowedNamespaces`: a list of namespaces that are permitted to either create new buckets or to access existing buckets.
+1. `protocol`: protocol-specific field the application will use to access the backend storage.
+     - `name`: supported protocols are: "s3", "gs", and "azureBlob".
+     - `version`: version being used by the backend storage.
+     - `azureBlob`: data required to target a provisioned azure container.
+     - `gs`: data required to target a provisioned GS bucket.
+     - `s3`: data required to target a provisioned S3 bucket.
+1. `parameters`: a copy of the BucketClass parameters.
```

Comment:
@wlan0 @msau42 reminded me that PV/PVC actually handles this in a better way. Instead of copying all parameters over from the Class object (which may include lots of unnecessary information), the CSI `CreateVolumeResponse` contains a set of `volume_attributes` -- basically letting the driver decide what additional important info should be maintained about the volume. We should do something similar here.